### DISCUSSION TITLE
How to add `alg` to generated JWK?

### DISCUSSION THREAD

**Discussion by ghost** [lestrrat-go/jwx#708] [General]
Hi,

I've used the samples in the documentation to generate a RSA JWKS. However, I see that `alg` is not set
```go
privateKey, err := rsa.GenerateKey(rand.Reader, keySize)
	if err != nil {
		return nil, err
	}
privateJWK := jwk.NewRSAPrivateKey()
err = privateJWK.FromRaw(privateKey)
	if err != nil {
		return nil, err
	}

publicJWK := jwk.NewRSAPublicKey()
	err = publicJWK.FromRaw(&privateKey.PublicKey)
	if err != nil {
		return nil, err
	}

fmt.Println(publicJWK.Algorithm()) // ""
```

How should I set the `alg` field? I feel like using `Set("alg","RS256")` is a bit brute-force, no?

Thanks for the help!

**Comment by lestrrat** [comment]
The answer _is_ to use `Set()`. You could use `jwk.AlgorithmKey` and `jwa.RS256` as constants, but that's about it.
What's making it feel like it "brute-force"? I'd like to know what seems to be the problem.

**Reply by phrozen** [reply]
Can I deduce the algorithm of a given key parsed `FromRaw()`? I think what he means as "bruteforce" is hardcoding the algorithm when creating a JWKS instead of relying on the key itself.

I know is a good practice to manually set it when verifying tokens, but when doing a server... how do I pick the right algorithm based on the input `key`?

**Reply by lestrrat** [reply]
You can't deduce _exactly_.
You can deduce that an EC key will use one of the ESXXX algorithms, but not necessarily which one. Same for RSA keys.

However, if you are the server, presumably you are the one who signed/encrypted the payload (or you should be the one who told the user which algorithm to use), and therefore you should know which algorithm was used.

At least that's what I think. Does that make sense?

**Reply by phrozen** [reply]
Yes it does. I'm writing a B2B server that supports only client credentials flow/grant, mean to be used for micro service auth. I plan to open source it once its done, I really don't want to tie users to specific algorithm(s), that is why I was looking for key algorithm introspection. I guess users will have to provide the algorithm along with the key in the configuration then...

**Reply by lestrrat** [reply]
I actually forgot that I had written `jws.AlgorithmForKey()`.  This will return the list of possible algorithms for a given key type, which you could try to match with the actual keys. This requires you to try each algorithm until you find a match, so if you can allow that behavior, you could do that too.

**Reply by gestrada-kr** [reply]
I can definitely try it! getting the algorithm for the key is a one time only operation every time the service starts or the keys are rotated, so I can definitely allow it. I will look into that one thanks!

**Reply by gestrada-kr** [reply]
@lestrrat what would the process to match the algorithm to the key would be like? For example, if I create a key with `rsa.GenerateKey(rand.Reader, 2048)` the output of `jws.AlgorithmForKey()` (after `FromRaw()`) is: `[RS256 RS384 RS512 PS256 PS384 PS512]` 

From there, how do I match it to the specific one?

**Reply by lestrrat** [reply]
brute force and try each one of them! :D There's just no other way.

---

As an aside, I guess my `AlgorithmsForKey` doesn't look at the actual Go key type (sorry, it has been a while)?
You might be able to patch AlgorithmsForKey to look at the actual key type to make slightly smarter guesses.

**Reply by gestrada-kr** [reply]
Thanks @lestrrat, I guess that if `AlgorithmForKeys()` is not used anywhere in the code for actual type determination, I might as well patch it, make it so that it can actually determine the algorithm and submit a PR. Does that sound good?

**Reply by lestrrat** [reply]
`AlgorithmForKeys()` is being used within `jws.WithKeySet()`  https://github.com/lestrrat-go/jwx/blob/develop/v2/jws/key_provider.go#L128

If you are filing a PR, please make sure to provide tests so we know exactly how you intend to use the new features.

**Comment by lestrrat** [comment]
Also, it _is_ in the docs ;)

https://github.com/lestrrat-go/jwx/blob/develop/v2/docs/01-jwt.md#parse-and-verify-a-jwt-with-a-key-set-matching-kid
https://github.com/lestrrat-go/jwx/blob/develop/v2/docs/02-jws.md#verification-using-a-jwks
https://github.com/lestrrat-go/jwx/blob/develop/v2/docs/03-jwe.md#decrypting-using-a-jwks

**Comment by ghost** [comment]
I guess I need to brush up on my JOSE practices 🙏
